crates/payload/basic/src/lib.rs

```diff
@@ -0,0 +1,561 @@
+#[derive(Debug, Clone)]
+pub struct BasicPayloadJobGeneratorConfig {
+    /// The interval at which the job should build a new payload after the last.
+    interval: Duration,
+    /// The deadline when this job should resolve.
+    deadline: Duration,
+    /// Maximum number of tasks to spawn for building a payload.
+    max_payload_tasks: usize,
+}
+
+// === impl BasicPayloadJobGeneratorConfig ===
+
+impl BasicPayloadJobGeneratorConfig {
+    /// Sets the interval at which the job should build a new payload after the last.
+    pub fn interval(mut self, interval: Duration) -> Self {
+        self.interval = interval;
+        self
+    }
+
+    /// Sets the deadline when this job should resolve.
+    pub fn deadline(mut self, deadline: Duration) -> Self {
+        self.deadline = deadline;
+        self
+    }
+
+    /// Sets the maximum number of tasks to spawn for building a payload(s).
+    ///
+    /// # Panics
+    ///
+    /// If `max_payload_tasks` is 0.
+    pub fn max_payload_tasks(mut self, max_payload_tasks: usize) -> Self {
+        assert!(max_payload_tasks > 0, "max_payload_tasks must be greater than 0");
+        self.max_payload_tasks = max_payload_tasks;
+        self
+    }
+}
+
+impl Default for BasicPayloadJobGeneratorConfig {
+    fn default() -> Self {
+        Self {
+            interval: Duration::from_secs(1),
+            // 12s slot time
+            deadline: Duration::from_secs(12),
+            max_payload_tasks: 3,
+        }
+    }
+}
+
+/// A basic payload job that continuously builds a payload with the best transactions from the pool.
+pub struct BasicPayloadJob<Client, Pool, Tasks> {
+    /// The configuration for how the payload will be created.
+    config: PayloadConfig,
+    /// The client that can interact with the chain.
+    client: Client,
+    /// The transaction pool.
+    pool: Pool,
+    /// How to spawn building tasks
+    executor: Tasks,
+    /// The deadline when this job should resolve.
+    deadline: Pin<Box<Sleep>>,
+    /// The interval at which the job should build a new payload after the last.
+    interval: Interval,
+    /// The best payload so far.
+    best_payload: Option<Arc<BuiltPayload>>,
+    /// Receiver for the block that is currently being built.
+    pending_block: Option<PendingPayload>,
+    /// Restricts how many generator tasks can be executed at once.
+    payload_task_guard: PayloadTaskGuard,
+}
+
+// === impl BasicPayloadJob ===
+
+impl<Client, Pool, Tasks> BasicPayloadJob<Client, Pool, Tasks> {
+    /// Checks if the new payload is better than the current best.
+    ///
+    /// This compares the total fees of the blocks, higher is better.
+    fn is_better(&self, new_payload: &BuiltPayload) -> bool {
+        if let Some(best_payload) = &self.best_payload {
+            new_payload.fees() > best_payload.fees()
+        } else {
+            true
+        }
+    }
+}
+
+impl<Client, Pool, Tasks> Stream for BasicPayloadJob<Client, Pool, Tasks>
```

Comment:
I like this pattern, each job is responsible for everything on its own, vs having some master task that handles things itself, the master task just drives the jobs a nd the complexity is encapsulated there